proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+TTH (i.e. the compiler may reject programs using TTH under
+``NoImplicitStagePersistence``).  In the long term, we believe that implementing
+Staging with Class is desirable and consistent with the direction of travel
+established by this proposal, but the full details of Staging with Class are out
+of scope.
+
+
+Implicit lifting and deriving ``Lift`` instances
+################################################
+.. _lift_instances:
+
+``Lift`` instances are used to provide serialisation-based cross-stage
+persistence.  For example, a typical ``Lift`` instance looks like::
+
+    data MInt = Some Int | None
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+The presence of this instance means the following declaration will be accepted::
+
+  foo :: MInt -> Q Exp
+  foo x = [| x |]  -- implicitly becomes [| $(lift x) |]
+
+Defining a ``Lift`` instance requires the datatype constructors to be available
+both at compile-time and runtime, so defining ``Lift`` within the same module as
+the datatype itself requires path-based cross-stage persistence.  Operationally,
+``None`` and ``Some`` are needed both at compile-time *and*  runtime since they
+are both matched on at compile time, and also persisted to be spliced in the
+future into a program that can make use of them at runtime. As a result, it
+isn't possible to define or derive a (non-orphan) ``Lift`` instance under
+``NoImplicitStagePersistence``.
+
+An orphan ``Lift`` instance can be defined thus::
+
+  module M where
+    data MInt = Some Int | None
+
+  module N where
+    import M
+    import quote M
+
+    instance Lift MInt where
+        lift :: MInt -> Q Exp
+        lift None     = [| None |]
+        lift (Some x) = [| Some $(lift x) |]
+
+This isn't technically problematic, rather it is just a result of what ``Lift``
+means. However, it means some users may need to modify their use of ``Lift``
+instances if they wish to benefit more from ``NoImplicitStagePersistence``.
+Users are free to use ``ImplicitStagePersistence`` in selected modules to allow
+defining ``Lift`` instances, but doing so means all the dependencies of the
+module will need to be available both at compile-time and runtime.
+
+As a general rule, ``Lift`` instances should be defined only for simple
+datatypes near the root of the module hierarchy of an application.
+
+Just as ``NoImplicitStagePersistence`` allows users to disable implicit
+path-based cross-stage persistence, it would make sense to have an extension
+flag to disable implicit lifting (serialisation-based persistence).  This would
+allow the programmer to ensure they are explicit about where calls to ``lift``
+occur in their programs, which is sometimes desirable when using staging for
+runtime performance.  We intend to bring forward a separate proposal for this,
+as it is otherwise orthogonal to the current proposal.
```

Comment:
TBH, upon reading the spec of `NoImplicitStagePersistence` I was immediately wondering why it would not also deactivate serialisation-based persistence. 

How would you name this orthogonal extension flag if `NoImplicitStagePersistence` is already taken?